staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go

```diff
@@ -205,18 +228,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 				APIGroupVersion: gv.String(),
 				APIResource:     resource,
 			})
+			apiResources = append(apiResources, resource)
+		}
+		if len(apiResources) > 0 {
+			apiList.APIResources = apiResources
+			filteredList = append(filteredList, apiList)
 		}
 	}
 
-	if o.NoHeaders == false && o.Output != "name" {
-		if err = printContextHeaders(w, o.Output); err != nil {
+	if !o.NoHeaders && (*o.PrintFlags.OutputFormat == "" || *o.PrintFlags.OutputFormat == "wide") {
+		if err = printContextHeaders(w, *o.PrintFlags.OutputFormat); err != nil {
 			return err
 		}
 	}
 
+	if *o.PrintFlags.OutputFormat == "json" || *o.PrintFlags.OutputFormat == "yaml" {
+		flatList := &metav1.APIResourceList{
```

Comment:
What is the reason you can't create this `flatList` already when creating `filteredList`? With appropriate comment above that variable.